src/ZkFold/Base/Algebra/Polynomials/Multivariate/Monomial.hs

```diff
@@ -31,47 +18,55 @@ import           Prelude                         hiding (Num (..), drop, filter,
 import           Test.QuickCheck                 (Arbitrary (..))
 
 import           ZkFold.Base.Algebra.Basic.Class
-import           ZkFold.Base.Data.Vector         (Vector (..))
-import           ZkFold.Prelude                  (elemIndex)
+import           ZkFold.Prelude                  (elemIndex, (!!))
 
-type Variable i = Ord i
+type Variable i = (Eq i, Ord i)
 
 type Monomial i j = (Variable i, Ord j, Semiring j)
 
 -- | Monomial type
-newtype M i j m = M m
+newtype Mono i j = M (Map i j)
     deriving (Generic, NFData, FromJSON, ToJSON)
 
--- | Most general type for a multivariate monomial
-type Monomial' = M Natural Natural (Map Natural Natural)
+------------------------------------ Map-based monomials ------------------------------------
 
-type MonomialRepAny = Map Integer Integer
+-- | Monomial constructor
+monomial :: Monomial i j => Map i j -> Mono i j
+monomial = M . filter (/= zero)
 
-type MonomialRepBoundedDegree i d = Vector d (i, Bool)
+dividable :: forall i j . Monomial i j => Mono i j -> Mono i j -> Bool
+dividable (M l) (M r) = isSubmapOfBy (<=) r l
 
-type MonomialAny = M Integer Integer MonomialRepAny
+evalMonomial :: forall i j b .
+    MultiplicativeMonoid b =>
+    Exponent b j =>
+    (i -> b) -> Mono i j -> b
+evalMonomial f (M m) =
+    foldrWithKey (\i j x -> (f i ^ j) * x) (one @b) m
 
-type MonomialBoundedDegree i d = M i Bool (MonomialRepBoundedDegree i d)
+mapVar :: Variable i => [i] -> [i] -> i -> i
```

Comment:
Could you please add a description of the meaning behind this function? 
Also, is there a way to make it return Maybe? I just don't like the runtime error

Reply:
I changed the function signature and added a comment. For now, I left the runtime error.